File: Source/Skeleton/SkeletonObj.c
/****************************/
/*   	SKELETON.C    	    */
/* (c)2003 Pangea Software  */
/* By Brian Greenstone      */
/****************************/


/****************************/
/*    EXTERNALS             */
/****************************/

#include "3DMath.h"

extern	EventRecord			gTheEvent;
extern	NewObjectDefinitionType	gNewObjectDefinition;
extern	OGLSetupOutputType		*gGameViewInfoPtr;


/****************************/
/*    PROTOTYPES            */
/****************************/

static SkeletonObjDataType *MakeNewSkeletonBaseData(short sourceSkeletonNum);
static void DisposeSkeletonDefinitionMemory(SkeletonDefType *skeleton);


/****************************/
/*    CONSTANTS             */
/****************************/


/*********************/
/*    VARIABLES      */
/*********************/

static SkeletonDefType		*gLoadedSkeletonsList[MAX_SKELETON_TYPES];

static short	    gNumDecomposedTriMeshesInSkeleton[MAX_SKELETON_TYPES];


/**************** INIT SKELETON MANAGER *********************/

void InitSkeletonManager(void)
{
short	i;

	CalcAccelerationSplineCurve();									// calc accel curve

	for (i =0; i < MAX_SKELETON_TYPES; i++)
		gLoadedSkeletonsList[i] = nil;

}


/******************** LOAD A SKELETON ****************************/

void LoadASkeleton(Byte num, OGLSetupOutputType *setupInfo)
{
	if (num >= MAX_SKELETON_TYPES)
		DoFatalAlert("\pLoadASkeleton: MAX_SKELETON_TYPES exceeded!");

	if (gLoadedSkeletonsList[num] != nil)								// check if already loaded
		DoFatalAlert("\pLoadASkeleton:  skeleton already loaded!");


				/* LOAD THE SKELETON FILE */

	gLoadedSkeletonsList[num] = LoadSkeletonFile(num, setupInfo);


	gNumDecomposedTriMeshesInSkeleton[num] = gLoadedSkeletonsList[num]->numDecomposedTriMeshes;		// keep easy access version of this value
}





/****************** FREE SKELETON FILE **************************/
//
// Disposes of all memory used by a skeleton file (from File.c)
//

void FreeSkeletonFile(Byte skeletonType)
{
short	i;

	if (gLoadedSkeletonsList[skeletonType])										// make sure this really exists
	{
		for (i=0; i < gNumDecomposedTriMeshesInSkeleton[skeletonType]; i++)		// dispose of the local copies of the decomposed trimeshes
		{
//			MO_DeleteObjectInfo_Geometry_VertexArray(&gLocalTriMeshesOfSkelType[skeletonType][i]);	// delete the data
		}

		DisposeSkeletonDefinitionMemory(gLoadedSkeletonsList[skeletonType]);	// free skeleton data
		gLoadedSkeletonsList[skeletonType] = nil;
	}
}


/*************** FREE ALL SKELETON FILES ***************************/
//
// Free's all except for the input type (-1 == none to skip)
//

void FreeAllSkeletonFiles(short skipMe)
{
short	i;

	for (i = 0; i < MAX_SKELETON_TYPES; i++)
	{
		if (i != skipMe)
	 		FreeSkeletonFile(i);
	}
}

#pragma mark -

/***************** MAKE NEW SKELETON OBJECT *******************/
//
// This routine simply initializes the blank object.
// The function CopySkeletonInfoToNewSkeleton actually attaches the specific skeleton
// file to this ObjNode.
//

ObjNode	*MakeNewSkeletonObject(NewObjectDefinitionType *newObjDef)
{
ObjNode	*newNode;
int		type;
float	scale;

	type = newObjDef->type;
	scale = newObjDef->scale;

			/* CREATE NEW OBJECT NODE */

	newObjDef->genre = SKELETON_GENRE;
	newNode = MakeNewObject(newObjDef);
	if (newNode == nil)
		return(nil);


			/* LOAD SKELETON FILE INTO OBJECT */

	newNode->Skeleton = MakeNewSkeletonBaseData(type); 			// alloc & set skeleton data
	if (newNode->Skeleton == nil)
		DoFatalAlert("\pMakeNewSkeletonObject: MakeNewSkeletonBaseData == nil");

	UpdateObjectTransforms(newNode);


			/*  SET INITIAL DEFAULT POSITION */

	SetSkeletonAnim(newNode->Skeleton, newObjDef->animNum);
	UpdateSkeletonAnimation(newNode);
	UpdateSkinnedGeometry(newNode);								// prime the trimesh

	CalcObjectRadiusFromBBox(newNode);							// set correct bounding sphere


	newNode->VertexArrayMode = VERTEX_ARRAY_RANGE_TYPE_SKELETONS;

	return(newNode);
}




/***************** ALLOC SKELETON DEFINITION MEMORY **********************/
//
// Allocates all of the sub-arrays for a skeleton file's definition data.
// ONLY called by ReadDataFromSkeletonFile in file.c.
//
// NOTE: skeleton has already been allocated by LoadSkeleton!!!
//

void AllocSkeletonDefinitionMemory(SkeletonDefType *skeleton)
{
long	numAnims,numJoints;

	numJoints = skeleton->NumBones;														// get # joints in skeleton
	numAnims = skeleton->NumAnims;														// get # anims in skeleton

				/***************************/
				/* ALLOC ANIM EVENTS LISTS */
				/***************************/

	skeleton->NumAnimEvents = (Byte *)AllocPtr(sizeof(Byte)*numAnims);					// array which holds # events for each anim
	Alloc_2d_array(AnimEventType, skeleton->AnimEventsList, numAnims, MAX_ANIM_EVENTS);


			/* ALLOC BONE INFO */

	skeleton->Bones = (BoneDefinitionType *)AllocPtr(sizeof(BoneDefinitionType)*numJoints);


		/* ALLOC DECOMPOSED DATA */

	skeleton->decomposedPointList = (DecomposedPointType *)AllocPtr(sizeof(DecomposedPointType)*MAX_DECOMPOSED_POINTS);
	skeleton->decomposedNormalsList = (OGLVector3D *)AllocPtr(sizeof(OGLVector3D)*MAX_DECOMPOSED_NORMALS);
}


/*************** DISPOSE SKELETON DEFINITION MEMORY ***************************/
//
// Disposes of all alloced memory (from above) used by a skeleton file definition.
//

static void DisposeSkeletonDefinitionMemory(SkeletonDefType *skeleton)
{
short	j,numAnims,numJoints;

	if (skeleton == nil)
		return;

	numAnims = skeleton->NumAnims;										// get # anims in skeleton
	numJoints = skeleton->NumBones;

			/* NUKE THE SKELETON BONE POINT & NORMAL INDEX ARRAYS */

	for (j=0; j < numJoints; j++)
	{
		if (skeleton->Bones[j].pointList)
			SafeDisposePtr(skeleton->Bones[j].pointList);
		if (skeleton->Bones[j].normalList)
			SafeDisposePtr(skeleton->Bones[j].normalList);
	}
	SafeDisposePtr((Ptr)skeleton->Bones);									// free bones array
	skeleton->Bones = nil;

				/* DISPOSE ANIM EVENTS LISTS */

	SafeDisposePtr((Ptr)skeleton->NumAnimEvents);

	Free_2d_array(skeleton->AnimEventsList);


			/* DISPOSE JOINT INFO */

	for (j=0; j < numJoints; j++)
	{
		Free_2d_array(skeleton->JointKeyframes[j].keyFrames);		// dispose 2D array of keyframe data

		skeleton->JointKeyframes[j].keyFrames = nil;
	}

			/* DISPOSE DECOMPOSED DATA ARRAYS */

//	for (j = 0; j < skeleton->numDecomposedTriMeshes; j++)			// first dispose of the trimesh data in there
//	{
//		MO_DisposeObject_Geometry_VertexArray(&skeleton->decomposedTriMeshes[j]);		// dispose of material refs
//		MO_DeleteObjectInfo_Geometry_VertexArray(&skeleton->decomposedTriMeshes[j]);	// free the arrays
//	}

	if (skeleton->decomposedPointList)
	{
		SafeDisposePtr((Ptr)skeleton->decomposedPointList);
		skeleton->decomposedPointList = nil;
	}

	if (skeleton->decomposedNormalsList)
	{
		SafeDisposePtr((Ptr)skeleton->decomposedNormalsList);
		skeleton->decomposedNormalsList = nil;
	}

			/* DISPOSE OF MASTER DEFINITION BLOCK */

	SafeDisposePtr((Ptr)skeleton);
}

#pragma mark -


/****************** MAKE NEW SKELETON OBJECT DATA *********************/
//
// Allocates & inits the Skeleton data for an ObjNode.
//

static SkeletonObjDataType *MakeNewSkeletonBaseData(short skeletonNum)
{
SkeletonDefType		*skeletonDefPtr;
SkeletonObjDataType	*skeletonData;
int					i, numDecomp;


	skeletonDefPtr = gLoadedSkeletonsList[skeletonNum];				// get ptr to source skeleton definition info
	if (skeletonDefPtr == nil)
	{
		DoAlert("\pMakeNewSkeletonBaseData: Skeleton data isnt loaded!");
		ShowSystemErr(skeletonNum);
	}


			/* ALLOC MEMORY FOR NEW SKELETON OBJECT DATA STRUCTURE */

	skeletonData = (SkeletonObjDataType *)AllocPtrClear(sizeof(SkeletonObjDataType));


			/* INIT NEW SKELETON */

	skeletonData->skeletonDefinition	= skeletonDefPtr;					// point to source skeletal data
	skeletonData->AnimSpeed 			= 1.0;
	skeletonData->JointsAreGlobal 		= false;

	for (i = 0; i < MAX_DECOMPOSED_TRIMESHES; i++)
		skeletonData->overrideTexture[i] = nil;								// assume no override texture.. yet.


			/****************************************/
			/* MAKE COPY OF TRIMESHES FOR LOCAL USE */
			/****************************************/
			//
			// Each ObjNode's Skeleton Def will have it's own triMesh copies
			// of the actual skeleton's geometry.  There are 2 copies because we
			// double-buffer it for VAR.  These local copies are what
			// we can modify to perform deformation animation on.
			//

	numDecomp = gNumDecomposedTriMeshesInSkeleton[skeletonNum];		// how many trimeshes in this skeleton's geometry?

	for (i=0; i < numDecomp; i++)
	{
		MO_DuplicateVertexArrayData(&gLoadedSkeletonsList[skeletonNum]->decomposedTriMeshes[i],
									&skeletonData->deformedMeshes[0][i], VERTEX_ARRAY_RANGE_TYPE_SKELETONS);

		MO_DuplicateVertexArrayData(&gLoadedSkeletonsList[skeletonNum]->decomposedTriMeshes[i],
									&skeletonData->deformedMeshes[1][i], VERTEX_ARRAY_RANGE_TYPE_SKELETONS+1);
	}


		/* CREATE AN OPENGL "FENCE" SO THAT WE CAN TELL WHEN DRAWING THE SKELETON IS DONE */
		//
		// NOTE:  We only use these fences for determinine when it is safe do delete a skeleton.
		// 			We do NOT use this fence for determining when it is safe to modify the geometry
		//			because we are using double-buffered vertex array data for each skeleton.
		//

	if (gUsingVertexArrayRange)
	{
		glGenFencesAPPLE(1, &skeletonData->oglFence);
		skeletonData->oglFenceIsActive = false;
	}


	return(skeletonData);
}


/************************ FREE SKELETON BASE DATA ***************************/

void FreeSkeletonBaseData(SkeletonObjDataType *skeletonData, short skeletonType)
{
short	numDecomp, i;

		/* WAIT FOR OPENGL FENCE MARKER BEFORE DELETING VAR'S */
		//
		// The geometry might still be in the GPU, so be sure it's finished before
		// deleting it.  Deleting it isn't bad, but if something else uses the same memory
		// on this same frame, then that could be bad.
		//

	if (gUsingVertexArrayRange)
	{
		if (skeletonData->oglFenceIsActive)
			glFinishFenceAPPLE(skeletonData->oglFence);

			/* DISPOSE OF THE OPENGL "FENCE" */

		glDeleteFencesAPPLE(1, &skeletonData->oglFence);
		skeletonData->oglFence = nil;
		skeletonData->oglFenceIsActive = false;
	}


		/* FREE OUR LOCAL COPY OF THE SKELETON'S TRIMESH */

	numDecomp = gNumDecomposedTriMeshesInSkeleton[skeletonType];					// how many trimeshes in this skeleton's geometry?

	for (i = 0; i < numDecomp; i++)
	{
		MO_DeleteObjectInfo_Geometry_VertexArray(&skeletonData->deformedMeshes[0][i]);		// free both double-buffers
		MO_DeleteObjectInfo_Geometry_VertexArray(&skeletonData->deformedMeshes[1][i]);
	}



		/* FREE THE SKELETON DATA STRUCT */

	SafeDisposePtr(skeletonData);
}



#pragma mark -

/*************************** DRAW SKELETON ******************************/

void DrawSkeleton(ObjNode *theNode, const OGLSetupOutputType *setupInfo)
{
short				i,numTriMeshes;
short				skelType;
MOMaterialObject	*overrideTexture, *oldTexture = nil;
SkeletonObjDataType	*skeleton = theNode->Skeleton;
Byte				buffNum;

	buffNum = setupInfo->frameCount & 1;

	skelType = theNode->Type;
	numTriMeshes = gNumDecomposedTriMeshesInSkeleton[skelType];						// get # trimeshes to draw

	for (i = 0; i < numTriMeshes; i++)												// submit each trimesh of it
	{
		MOVertexArrayData *triMesh = &skeleton->deformedMeshes[buffNum][i];			// point to triMesh

		overrideTexture = skeleton->overrideTexture[i];								// get any override texture ref (illegal ref)
		if (overrideTexture)														// set override texture
		{
			if (triMesh->numMaterials > 0)
			{
				oldTexture = triMesh->materials[0];									// get the real texture for this mesh
				triMesh->materials[0] = overrideTexture;							// set the override texture (temporarily)
			}
		}

					/* SUBMIT IT */

		MO_DrawGeometry_VertexArray(triMesh, setupInfo);


		if (overrideTexture && oldTexture)											// see if need to set texture back to normal
			triMesh->materials[0] = oldTexture;
	}


		/* INSERT AN OPENGL "FENCE" INTO THE DATA STREAM SO THAT WE CAN DETECT WHEN DRAWING THIS SKELETON IS COMPLETE */

	if (gCurrentSplitScreenPane == (gNumPlayers-1))				// only submit the fence on the last pane to be drawn
	{
		glSetFenceAPPLE(skeleton->oglFence);
		skeleton->oglFenceIsActive = true;						// it's ok to call glTestFenceAPPLE or glFinishFenceAPPLE now that we've set it
	}

}

File: Source/Skeleton/SkeletonObj.swift
// SkeletonObj.swift - Port of SkeletonObj.c to Swift
//
// VERTEXARRAYRANGES is hardcoded to 0 in game.h, so the #if VERTEXARRAYRANGES
// blocks in the original C are dead code in every build this port targets
// (same reasoning as #if SKIPFLUFF elsewhere) — dropped here.

private var gLoadedSkeletonsList: InlineArray<7, UnsafeMutablePointer<SkeletonDefType>?> = InlineArray(repeating: nil)
private var gNumDecomposedTriMeshesInSkeleton: InlineArray<7, Int16> = InlineArray(repeating: 0)

func InitSkeletonManager() {
    CalcAccelerationSplineCurve() // calc accel curve

    for (i, _) in SkeletonType.allCases.enumerated() {
        gLoadedSkeletonsList[i] = nil
    }
}

func LoadASkeleton(_ num: UInt8) {
    if num >= UInt8(SkeletonType.allCases.count) {
        SwFatal("LoadASkeleton: MAX_SKELETON_TYPES exceeded!")
    }

    if gLoadedSkeletonsList[Int(num)] != nil { // check if already loaded
        SwFatal("LoadASkeleton:  skeleton already loaded!")
    }

    // LOAD THE SKELETON FILE

    let loaded = LoadSkeletonFile(Int16(num))!
    gLoadedSkeletonsList[Int(num)] = loaded

    gNumDecomposedTriMeshesInSkeleton[Int(num)] = Int16(loaded.pointee.numDecomposedTriMeshes) // keep easy access version of this value
}

// Disposes of all memory used by a skeleton file (from File.c)
func FreeSkeletonFile(_ skeletonType: UInt8) {
    if let skeleton = gLoadedSkeletonsList[Int(skeletonType)] { // make sure this really exists
        // (the local-copy-of-decomposed-trimeshes disposal loop was already
        // commented out in the original C)

        disposeSkeletonDefinitionMemory(skeleton) // free skeleton data
        gLoadedSkeletonsList[Int(skeletonType)] = nil
    }
}

// Free's all except for the input type (-1 == none to skip)
func FreeAllSkeletonFiles(_ skipMe: Int16) {
    for (i, _) in SkeletonType.allCases.enumerated() {
        if Int16(i) != skipMe {
            FreeSkeletonFile(UInt8(i))
        }
    }
}

// This routine simply initializes the blank object.
// The function CopySkeletonInfoToNewSkeleton actually attaches the specific skeleton
// file to this ObjNode.
func MakeNewSkeletonObject(_ newObjDef: UnsafeMutablePointer<NewObjectDefinitionType>!) -> UnsafeMutablePointer<ObjNode>! {
    let type = newObjDef.pointee.type

    // CREATE NEW OBJECT NODE

    newObjDef.pointee.genre = UInt8(SKELETON_GENRE)
    guard let newNode = MakeNewObject(newObjDef) else {
        return nil
    }

    // LOAD SKELETON FILE INTO OBJECT

    newNode.pointee.Skeleton = makeNewSkeletonBaseData(Int16(type)) // alloc & set skeleton data
    if newNode.pointee.Skeleton == nil {
        SwFatal("MakeNewSkeletonObject: MakeNewSkeletonBaseData == nil")
    }

    newNode.updateTransforms()

    // SET INITIAL DEFAULT POSITION

    SetSkeletonAnim(newNode.pointee.Skeleton, Int(newObjDef.pointee.animNum))
    UpdateSkeletonAnimation(newNode)
    UpdateSkinnedGeometry(newNode) // prime the trimesh

    newNode.calcRadiusFromBBox() // set correct bounding sphere

    newNode.pointee.VertexArrayMode = UInt8(VertexArrayRangeType.skeletons.rawValue)

    return newNode
}

// Allocates all of the sub-arrays for a skeleton file's definition data.
// ONLY called by ReadDataFromSkeletonFile in file.c.
//
// NOTE: skeleton has already been allocated by LoadSkeleton!!!
func AllocSkeletonDefinitionMemory(_ skeleton: UnsafeMutablePointer<SkeletonDefType>!) {
    let numJoints = Int(skeleton.pointee.NumBones) // get # joints in skeleton
    let numAnims = Int(skeleton.pointee.NumAnims) // get # anims in skeleton

    // ALLOC ANIM EVENTS LISTS

    skeleton.pointee.NumAnimEvents = AllocPtrClear(MemoryLayout<UInt8>.size * numAnims)!.assumingMemoryBound(to: UInt8.self) // array which holds # events for each anim
    skeleton.pointee.AnimEventsList = alloc2DArray(AnimEventType.self, rows: numAnims, cols: Int(MAX_ANIM_EVENTS))

    // ALLOC BONE INFO

    skeleton.pointee.Bones = AllocPtrClear(MemoryLayout<BoneDefinitionType>.size * numJoints)!.assumingMemoryBound(to: BoneDefinitionType.self)

    // ALLOC DECOMPOSED DATA

    skeleton.pointee.decomposedPointList = AllocPtrClear(MemoryLayout<DecomposedPointType>.size * Int(MAX_DECOMPOSED_POINTS))!.assumingMemoryBound(to: DecomposedPointType.self)
    skeleton.pointee.decomposedNormalsList = AllocPtrClear(MemoryLayout<OGLVector3D>.size * Int(MAX_DECOMPOSED_NORMALS))!.assumingMemoryBound(to: OGLVector3D.self)
}

// Disposes of all alloced memory (from above) used by a skeleton file definition.
private func disposeSkeletonDefinitionMemory(_ skeleton: UnsafeMutablePointer<SkeletonDefType>?) {
    guard let skeleton else {
        return
    }

    let numJoints = Int(skeleton.pointee.NumBones)

    // NUKE THE SKELETON BONE POINT & NORMAL INDEX ARRAYS

    for j in 0..<numJoints {
        if let pointList = skeleton.pointee.Bones![j].pointList {
            SafeDisposePtr(pointList)
        }
        if let normalList = skeleton.pointee.Bones![j].normalList {
            SafeDisposePtr(normalList)
        }
    }
    SafeDisposePtr(skeleton.pointee.Bones) // free bones array
    skeleton.pointee.Bones = nil

    // DISPOSE ANIM EVENTS LISTS

    SafeDisposePtr(skeleton.pointee.NumAnimEvents)

    free2DArray(skeleton.pointee.AnimEventsList)

    // DISPOSE JOINT INFO

    let jointKeyframesBase = UnsafeMutableRawPointer(skeleton.pointer(to: \.JointKeyframes)!).assumingMemoryBound(to: JointKeyFrameHeader.self)
    for j in 0..<numJoints {
        free2DArray(jointKeyframesBase[j].keyFrames) // dispose 2D array of keyframe data
        jointKeyframesBase[j].keyFrames = nil
    }

    // DISPOSE DECOMPOSED DATA ARRAYS
    //
    // (the trimesh-data disposal loop was already commented out in the original C)

    if let decomposedPointList = skeleton.pointee.decomposedPointList {
        SafeDisposePtr(decomposedPointList)
        skeleton.pointee.decomposedPointList = nil
    }

    if let decomposedNormalsList = skeleton.pointee.decomposedNormalsList {
        SafeDisposePtr(decomposedNormalsList)
        skeleton.pointee.decomposedNormalsList = nil
    }

    // DISPOSE OF MASTER DEFINITION BLOCK

    SafeDisposePtr(skeleton)
}

// Allocates & inits the Skeleton data for an ObjNode.
private func makeNewSkeletonBaseData(_ skeletonNum: Int16) -> UnsafeMutablePointer<SkeletonObjDataType>! {
    guard let skeletonDefPtr = gLoadedSkeletonsList[Int(skeletonNum)] else { // get ptr to source skeleton definition info
        SwFatal("MakeNewSkeletonBaseData: Skeleton data isn't loaded!")
        return nil
    }

    // ALLOC MEMORY FOR NEW SKELETON OBJECT DATA STRUCTURE

    let skeletonData = AllocPtrClear(MemoryLayout<SkeletonObjDataType>.size)!.assumingMemoryBound(to: SkeletonObjDataType.self)

    // INIT NEW SKELETON

    skeletonData.pointee.skeletonDefinition = skeletonDefPtr // point to source skeletal data
    skeletonData.pointee.AnimSpeed = 1.0
    skeletonData.jointsAreGlobal = false

    let overrideTextureBase = UnsafeMutableRawPointer(skeletonData.pointer(to: \.overrideTexture)!).assumingMemoryBound(to: UnsafeMutablePointer<MOMaterialObject>?.self)
    for i in 0..<Int(MAX_DECOMPOSED_TRIMESHES) {
        overrideTextureBase[i] = nil // assume no override texture.. yet.
    }

    // MAKE COPY OF TRIMESHES FOR LOCAL USE
    //
    // Each ObjNode's Skeleton Def will have it's own triMesh copies
    // of the actual skeleton's geometry.  There are 2 copies because we
    // double-buffer it for VAR.  These local copies are what
    // we can modify to perform deformation animation on.

    let numDecomp = Int(gNumDecomposedTriMeshesInSkeleton[Int(skeletonNum)]) // how many trimeshes in this skeleton's geometry?

    let srcTriMeshesBase = UnsafeMutableRawPointer(skeletonDefPtr.pointer(to: \.decomposedTriMeshes)!).assumingMemoryBound(to: MOVertexArrayData.self)
    let deformedMeshesBase = UnsafeMutableRawPointer(skeletonData.pointer(to: \.deformedMeshes)!).assumingMemoryBound(to: MOVertexArrayData.self)
    let deformedMeshesStride = Int(MAX_DECOMPOSED_TRIMESHES)

    for i in 0..<numDecomp {
        MO_DuplicateVertexArrayData(srcTriMeshesBase + i, deformedMeshesBase + (0 * deformedMeshesStride + i), Int16(VertexArrayRangeType.skeletons.rawValue))
        MO_DuplicateVertexArrayData(srcTriMeshesBase + i, deformedMeshesBase + (1 * deformedMeshesStride + i), Int16(VertexArrayRangeType.skeletons.rawValue) + 1)
    }

    return skeletonData
}

func FreeSkeletonBaseData(_ skeletonData: UnsafeMutablePointer<SkeletonObjDataType>!, _ skeletonType: Int16) {
    // FREE OUR LOCAL COPY OF THE SKELETON'S TRIMESH

    let numDecomp = Int(gNumDecomposedTriMeshesInSkeleton[Int(skeletonType)]) // how many trimeshes in this skeleton's geometry?

    let deformedMeshesBase = UnsafeMutableRawPointer(skeletonData.pointer(to: \.deformedMeshes)!).assumingMemoryBound(to: MOVertexArrayData.self)
    let deformedMeshesStride = Int(MAX_DECOMPOSED_TRIMESHES)

    for i in 0..<numDecomp {
        MO_DeleteObjectInfo_Geometry_VertexArray(deformedMeshesBase + (0 * deformedMeshesStride + i)) // free both double-buffers
        MO_DeleteObjectInfo_Geometry_VertexArray(deformedMeshesBase + (1 * deformedMeshesStride + i))
    }

    // FREE THE SKELETON DATA STRUCT

    SafeDisposePtr(skeletonData)
}

func DrawSkeleton(_ theNode: UnsafeMutablePointer<ObjNode>!) {
    let buffNum = Int(gGameViewInfoPtr!.pointee.frameCount & 1)

    let skelType = Int(theNode.pointee.Type)
    let numTriMeshes = Int(gNumDecomposedTriMeshesInSkeleton[skelType]) // get # trimeshes to draw

    let skeleton = theNode.pointee.Skeleton!
    let overrideTextureBase = UnsafeMutableRawPointer(skeleton.pointer(to: \.overrideTexture)!).assumingMemoryBound(to: UnsafeMutablePointer<MOMaterialObject>?.self)
    let deformedMeshesBase = UnsafeMutableRawPointer(skeleton.pointer(to: \.deformedMeshes)!).assumingMemoryBound(to: MOVertexArrayData.self)
    let deformedMeshesStride = Int(MAX_DECOMPOSED_TRIMESHES)

    for i in 0..<numTriMeshes { // submit each trimesh of it
        let triMesh = deformedMeshesBase + (buffNum * deformedMeshesStride + i) // point to triMesh

        let overrideTexture = overrideTextureBase[i] // get any override texture ref (illegal ref)
        var oldTexture: UnsafeMutablePointer<MOMaterialObject>?
        if let overrideTexture { // set override texture
            if triMesh.pointee.numMaterials > 0 {
                oldTexture = triMesh.pointee.materials.0 // get the real texture for this mesh
                triMesh.pointee.materials.0 = overrideTexture // set the override texture (temporarily)
            }
        }

        // SUBMIT IT

        MO_DrawGeometry_VertexArray(triMesh)

        if overrideTexture != nil, let oldTexture { // see if need to set texture back to normal
            triMesh.pointee.materials.0 = oldTexture
        }
    }
}
